src/github.com/matrix-org/dendrite/roomserver/query/query.go

```diff
@@ -374,58 +375,85 @@ func (r *RoomserverQueryAPI) QueryServerAllowedToSeeEvent(
 	ctx context.Context,
 	request *api.QueryServerAllowedToSeeEventRequest,
 	response *api.QueryServerAllowedToSeeEventResponse,
-) error {
-	stateEntries, err := state.LoadStateAtEvent(ctx, r.DB, request.EventID)
+) (err error) {
+	response.AllowedToSeeEvent, err = r.checkServerAllowedToSeeEvent(
+		ctx, request.EventID, request.ServerName,
+	)
+	return
+}
+
+func (r *RoomserverQueryAPI) checkServerAllowedToSeeEvent(
+	ctx context.Context, eventID string, serverName gomatrixserverlib.ServerName,
+) (bool, error) {
+	stateEntries, err := state.LoadStateAtEvent(ctx, r.DB, eventID)
 	if err != nil {
-		return err
+		return false, err
 	}
 
 	// TODO: We probably want to make it so that we don't have to pull
 	// out all the state if possible.
 	stateAtEvent, err := r.loadStateEvents(ctx, stateEntries)
 	if err != nil {
-		return err
+		return false, err
 	}
 
-	// TODO: Should this be lifted out of here to a more general set of
-	// auth functions?
-
-	isInRoom := false
-	for _, ev := range stateAtEvent {
-		membership, err := ev.Membership()
-		if err != nil {
-			continue
-		}
+	return auth.IsServerAllowed(serverName, stateAtEvent), nil
+}
 
-		if membership != "join" {
-			continue
-		}
+// QueryMissingEvents implements api.RoomserverQueryAPI
+func (r *RoomserverQueryAPI) QueryMissingEvents(
+	ctx context.Context,
+	request *api.QueryMissingEventsRequest,
+	response *api.QueryMissingEventsResponse,
+) error {
+	resultNIDs := make([]types.EventNID, 0, request.Limit)
+	var front []string
+	visited := make(map[string]bool, request.Limit) // request.Limit acts as a hint to size.
+	for _, id := range request.EarliestEvents {
+		visited[id] = true
+	}
 
-		stateKey := ev.StateKey()
-		if stateKey == nil {
-			continue
+	for _, id := range request.LatestEvents {
+		if !visited[id] {
+			front = append(front, id)
 		}
+	}
 
-		_, domain, err := gomatrixserverlib.SplitID('@', *stateKey)
+BFSLoop:
+	for len(front) > 0 {
+		var next []string
+		events, err := r.DB.EventsFromIDs(ctx, front)
 		if err != nil {
-			continue
+			return err
 		}
 
-		if domain == request.ServerName {
-			isInRoom = true
-			break
+		for _, ev := range events {
+			if len(resultNIDs) > request.Limit {
+				break BFSLoop
+			}
+			resultNIDs = append(resultNIDs, ev.EventNID)
+			for _, pre := range ev.PrevEventIDs() {
+				if !visited[pre] {
+					visited[pre] = true
+					allowed, err := r.checkServerAllowedToSeeEvent(
+						ctx, ev.EventID(), request.ServerName,
+					)
+					if err != nil {
+						return err
+					}
+
+					if allowed {
+						next = append(next, pre)
+					}
+				}
+			}
 		}
+		front = next
 	}
 
-	if isInRoom {
-		response.AllowedToSeeEvent = true
-		return nil
-	}
-
-	// TODO: Check if history visibility is shared and if the server is currently in the room
-
-	response.AllowedToSeeEvent = false
-	return nil
+	var err error
+	response.Events, err = r.loadEvents(ctx, resultNIDs)
```

Comment:
Can remove the `var err error` line and just change `... err = r.loadEvents` to `... err := r.loadEvents`.

Reply:
`response.Events` is a declared variable. :/

Reply:
Oh, turns out you can't `:=` assign to non-identifiers. My bad :p